### ISSUE TITLE
Standardisation of HTTP Status codes for Responses when posting to Inboxes

### ISSUE THREAD

**Issue by ThisIsMissEm** [w3c-social/activitypump#414]
Currently when a server wants to send something to another server, there is no standardisation as to what status codes an inbox should return for different conditions, so for instance, to determine success vs failure vs retryable, mastodon uses [the following](https://github.com/mastodon/mastodon/blob/main/app/helpers/jsonld_helper.rb#L198):

```ruby
  def response_successful?(response)
    # if the status code is >= 200 < 300
    (200...300).cover?(response.code)
  end

  def response_error_unsalvageable?(response)
    # status code is 501 or 4xx but not 401, 408, or 429.
    response.code == 501 || ((400...500).cover?(response.code) && ![401, 408, 429].include?(response.code))
  end
```

Other software likely has other implementations, for instance, for BookWyrm, if it receives an activity that it doesn't support, it currently returns a 404, when arguably a 400 Bad Request or similar status code would probably be better.

This leaves us in a situation where one server does not know whether to retry delivery or not, or even if the delivery was successful or not. General logic says "use the most appropriate http status code" but that creates ambiguity between implementations.

I think we would benefit from more strongly specifying status codes for inbox requests, as to indicate to the sending server that delivery was successful, or should be retried, or not retried.

**Comment by silverpill** [comment]
It would be great to define `410 Gone` as "please don't send anything to this inbox ever again, the actor was deleted"

**Comment by TallTed** [comment]
> It would be great to define `410 Gone` as "please don't send anything to this inbox ever again, the actor was deleted"

That seems too extreme, unless there is a rule that the identifier for such an actor/inbox can never be reused, even if control of the domain/identifier changes. So far as I know, such a rule does not exist, nor would it be implementable if it did.

So I think that "please don't send anything to this inbox ever again, the actor was deleted" is not viable. "Please delete this inbox from all current distributions; the actor was deleted," on the other hand, has some potential.

**Comment by snarfed** [comment]
@TallTed true! Reusing server domains and actor ids over time is a fascinating question more broadly. I wonder if anyone has ever thought it through or written anything up. Best practice for similar services like email and centralized social networks is to prevent reuse, since it can allow for impersonation, taking over accounts via email password resets, etc. But you're right, there's no concrete mechanism to prevent it across the board. Even keys and HTTP Sigs don't do it, since it's not easily distinguishable from key rotation, and right now we generally have to trust whatever key the domain serves for a given actor.

**Comment by TallTed** [comment]
> [@snarfed] Best practice for similar services like email and centralized social networks is to prevent reuse, since it can allow for impersonation, taking over accounts via email password resets, etc.

Believing that any identifier can only ever validly point to one entity is a problem. CoolURIs might not change, but that can only be held possible until the domain changes hands.

I have a long American name, first middle last and a Jr to round it out. Imagine my surprise when I received a mail delivery while working a temporary job, which was intended for someone at least 10 years my senior, who had exactly the same name! It was a bigger surprise for him, and for the sender of that mail, as its correct delivery was then delayed by some days, while it went back to the hub and thence out the proper spokes to reach him.

URIs are *better* identifiers in many ways than their predecessors or current alternatives, but they are far from perfect.

**Comment by evanp** [comment]
I think this is an excellent idea. This might be a good candidate for a FEP or for a CG report, depending on how core you think this issue is. (I think it's pretty core!)

As a first step, I'm going to create a primer page at https://www.w3.org/wiki/ActivityPub/Primer/HTTP_status_codes_for_delivery to start documenting this behaviour. That might be good input to a future standardisation effort.

**Comment by evanp** [comment]
I got through the 100s, 200s, 300s and 400s in the issue triage session on Jan 17, and will try to get the 500s finished next time around.

**Comment by silverpill** [comment]
@evanp Thank you for writing this page, it is very useful.

I agree with @TallTed the my initial suggestion regarding `410 Gone` was too extreme. Marking actor as "deleted" but allowing ID to be re-used would be preferable.

The description of `404 Not Found` should probably warn implementers that this status may indicate temporary unavailability due to server misconfiguration or software update, and should not be treated in the same way as `410 Gone`.

**Comment by ThisIsMissEm** [comment]
@evanp primer page now looks to be written, are there next steps we want to take here?
